### PR TITLE
docs: testing guidelines

### PR DESCRIPTION
**Pull request by robert-zaremba**
## Description

Adding testing guidelines
+ highlight the patterns and tools we should use more
+ highlight anti-patterns or practices we should use less
+ list tools which we currently use in the ecosystem.

The follow up will be to integrate mockgen with isolated integration test.

### CODE REVIEW

File: docs/testing.md [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+End-to-end test requires the test to setup the system components as they are in production. Real database, services, queues, etc. The chain network (with all required apps, like price-feeder) is setup using [dockertest](https://github.com/ory/dockertest) embedding actual binaries.
+
+These tests execute high level flow by broadcasting transactions to a node port or querying chain directly through the RPC.
+
+E2E tests live in `tests/e2e` package.
+
+## QA tests
+
+For production grade releases we need to have a testable backstage environment with some reliable testing data, accounts and tooling. At the same time we would like to easy functional tests and blockchain interaction with whole team: both for examples presentation, as well as quick and easy data load by a developer to play and smoke test the blockchain.
```
Comment (zarazan) [inline docs/testing.md:65]:
```suggestion
For production grade releases we need to have a testable backstage environment with some reliable testing data, accounts and tooling. At the same time we would like programmatic functional tests that run queries and transactions against the backstage environment that can be used for examples presentation, load tests, and smoke tests.
```

File: docs/testing.md [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+Simulation tests are defined in each module `simulations` package.
+
+## E2E tests
+
+End-to-end test requires the test to setup the system components as they are in production. Real database, services, queues, etc. The chain network (with all required apps, like price-feeder) is setup using [dockertest](https://github.com/ory/dockertest) embedding actual binaries.
+
+These tests execute high level flow by broadcasting transactions to a node port or querying chain directly through the RPC.
+
+E2E tests live in `tests/e2e` package.
+
```
Comment (zarazan) [inline docs/testing.md:62]:
```suggestion

E2E tests that run for a long time should be skipped in the CI that runs every PR and automatically ran every time a release is tagged, or every night.

```

File: docs/testing.md [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+## Simulation Tests
+
+Simulations are very high level App integration tests, where:
+
+- we only run App in it's expected setup (without the normally expected processes, like price-feeder)
+- App is producing blocks on the chain until a specified height is reached.
+  at each block the sim test runner sequentially ask each module to generate random transactions (represented as operations) and execute them directly in the app transaction runner.
+- during the sim runner, we don't execute any user defined assertions. Instead we expect the chain to function correctly (don't return errors nor panics). Hence, when creating a new operation (random sim transaction), we need to check conditions to satisfy that a transaction won't be rejected (like checking bank balance etc...).
+
```
Comment (zarazan) [inline docs/testing.md:52]: Adding this link under the simulation section could be helpful: https://pkg.go.dev/github.com/cosmos/cosmos-sdk/x/simulation#SimulateFromSeed

**Review by zarazan — APPROVED** [review]
Thanks for incorporating some of the info from the doc I made! I added a couple suggestions.

File: docs/testing.md [hunk 4]
```diff
@@ -0,0 +1,84 @@
+# Umee Testing Guildelines
+
+Rules of thumb:
+
+1. Design packages to make it easy to unit-test.
+   - Unit tests should represent majority of the test.
+   - Unit tests can be in the same package (to be able to test private methods) → no need to move unit tests to `_test` pacakge.
+2. Integration tests should be in `_test` package.
+3. Reduce tight coupling (will make it ) in the software.
```
[the remark this answers is not in the document]
Comment (robert-zaremba) [reply docs/testing.md:9]: its expanded in the section below. You should have join the workshop ;)

File: docs/testing.md (hunk 3 above)
Comment (robert-zaremba) [reply docs/testing.md:52]: added

**Pull request merged**